ML-Agents Code Restructure

dericp opened the pull request:
## Summary

The Python code in ML-Agents suffers from a package hierarchy that makes testing and publishing to PIP difficult. I propose that moving the `unityagents` and `unitytrainers` packages into a single top level `mlagents` python package will make the separation of these packages more clear, allow us to decouple their tests, and make it possible to publish to pip easily.

## Background

While working on moving curriculum learning out of unityagents and into unitytrainers, I wanted to be able to make a test hierarchy. Specifically, I wanted to create two separate test directories tests/unityagents and tests/unitytrainers.

Because of the way the Python code is currently packaged, this was difficult to do. Imports broke, and eventually the only way to get any tests to pass was to run `python -m pytest tests/` which is less than ideal.

Initially, we concluded that it would probably be a better use of my time to work on something else. However, with Arthur adding a gym environment to the repo, it seems that it might be better to perform a restructure as soon as possible.

## Goals

- Run all python tests from top level `ml-agents` directory by just running `$ pytest`.
- The `tests` package should mirror the structure of the Python code.
- Name the Python packages more intuitively.
- Publish `mlagents` package to PyPi.

## Plan

- Move directories around according to the new packaging scheme.
- Fix imports and tests.
- Expose learn.py as an environment script.
- Update documentation to reflect the new package structure.
- Update Docker container to reflect the new package structure.
- Upload `mlagents` to TestPyPi.
- Upload `mlagents` to PyPi.

## Things done that were not in original plan

- Added ml-agents-protobuf to the ml-agents repo. Documentation has been improved.
- Documentation that I had to edit has been changed to be compliant with markdownlint.

## Notes

There are a few work-items I could use help with:
- Verify that the Docker container is working properly.
  - I have been able to build and train an agent in the new Docker container but it would be great if @eshvk could sanity check my work.
- Verify that the `gym-unity` wrapper still works properly. @awjuliani 
- Verify that the "Training on AWS" documentation is correct.
- Verify that all behavior has remained identical to before.
  - All unit tests are passing but this does not necessarily give me confidence that all the code is working as it should. I guess we could either write more unit tests or empirically verify the code's correctness.
- Verify that all documentation is aligned with the new structure.

## Tips for Reviewing

This is a huge PR mostly because of edits in the documentation. I would suggest reviewing one commit at a time since each commit is a logical chunk of work.

To install `mlagents` from TestPyPi, run
```
pip install --index-url https://test.pypi.org/simple/ mlagents
```

vincentpierre commented on docs/Basic-Guide.md:
```diff
@@ -1,105 +1,154 @@
 # Basic Guide
```
It seems a lot of these changes are due to formatting of the markdowns. Is it possible to reduce the changes made to the docs and only include text changes ?

dericp replied:
Hmm, since the changes are already pushed it could be difficult. I figured that since we'd want to clean up the docs eventually I'd just do it at the same time. I agree it makes reviewing more difficult. If you feel strongly about undoing these changes, I'm happy to do it!

vincentpierre commented on docs/FAQ.md:
```diff
@@ -1,91 +1,106 @@
+communication port {} is still in use. "`, you can change the worker number in
+the Python script when calling
 
 `UnityEnvironment(file_name=filename, worker_id=X)`
```
This is python code

dericp replied:
Cool, changed the docs to highlight it as Python.

vincentpierre commented on docs/Python-API.md:
````diff
@@ -1,77 +1,136 @@
+environment is 3DBall.app, in python, run:
 
 ```python
 from unityagents import UnityEnvironment
````
`from mlagents.env`

dericp replied:
Good catch, fixed.

vincentpierre commented on ml-agents-protobuf/make.bat:
```diff
@@ -0,0 +1,43 @@
+# GRPC-TOOLS required. Install with `nuget install Grpc.Tools`. 
+# Then un-comment and replace [DIRECTORY] with location of files.
+# For example, on macOS, you might have something like:
+COMPILER=Grpc.Tools.1.14.1/tools/macosx_x64
```
Maybe let the default be `[Directory]` ?

dericp replied:
Woops, meant to comment that line out. I've gone ahead and done it.

vincentpierre commented on ml-agents-protobuf/make.bat:
```diff
@@ -0,0 +1,43 @@
+# COMPILER=[DIRECTORY]
+
+SRC_DIR=proto/mlagents/envs/communicator_objects
+DST_DIR_C=cs/CommunicatorObjects
```
Is there a way to generate directly into the appropriate folders ?

dericp replied:
Yep, now it automatically puts the proto files into the right folders.

vincentpierre commented on mlagents/envs/socket_communicator.py:
```diff
@@ -3,12 +3,12 @@
 
 logging.basicConfig(level=logging.INFO)
-logger = logging.getLogger("unityagents")
+logger = logging.getLogger("mlagents.envs")
```
Not sure what the logger should be, but `mlagents.envs` is too specific. We can spell out `ML-Agents` unless we want a different logger for each package (I am also okay with that)

dericp replied:
I was thinking we could have a different logger for `mlagents.envs` and `mlagents.trainers` which are the two main packages we ship.

eshvk commented:
Docker stuff LGTM

awjuliani commented:
Going to add comments here as I go through the repo.

First comment: the `ml-agents-protobuf` folder should just be called `protobuf`.

awjuliani commented:
I think this folder: https://github.com/Unity-Technologies/ml-agents/tree/develop-flat-code-restructure/python/mlagents should be called `ml-agents,` to distinguish it from the package itself.

awjuliani commented:
We will need to make a Readme for the ml-agents package.

awjuliani commented:
This file: https://github.com/Unity-Technologies/ml-agents/blob/develop-flat-code-restructure/python/mlagents/setup.py has a bunch of un-necessary comments.

awjuliani commented:
It is still unclear to me if we want `trainer-config` and `curricula` to live on the high-level of the repo. Perhaps they should go into `unity-volume`?

awjuliani commented on python/gym-unity/setup.py:
```diff
@@ -10,5 +10,5 @@
       author_email='ML-Agents@unity3d.com',
       url='https://github.com/Unity-Technologies/ml-agents',
       packages=find_packages(),
-      install_requires = ['gym', 'unityagents']
```
Technically this only needs to be 'mlagents[envs]`.

dericp replied:
To be clear, it doesn't require `gym`?

awjuliani replied:
It does! Just not `mlagents.trainers`.

awjuliani commented:
Another question is whether we want to split up the C# plugin from the environments here as well.
@vincentpierre did you test the symlinks on Windows?

dericp commented:
@awjuliani 

### Resolved

- Renamed `ml-agents-protobuf` to `protobuf-definitions`.
- Renamed `python/mlagents` to `python/ml-agents`.
- Removed comments from setup.py.

### Unresolved

- The `mlagents` package currently just uses the top level README.md file. Is that okay? I think it was like this before.
- I think it's fine for `trainer_config.yaml` and `curricula` to be at the top level. We could perhaps create a config folder for those files? I don't think `unity-volume` should be the place for config files since it's just a placeholder to demonstrate how to attach a docker volume, although it doesn't really do that and could probably be deleted.
- I think we decided that separating the environments and plugins was out of the scope of this refactor.

awjuliani commented:
@dericp Thanks for making all these changes. They all look great to me, except that you mis-spelled `definitions`.

dericp commented:
@awjuliani @vincentpierre Ready for merge. Will go ahead and do it as soon as you guys give me the ✅.

vincentpierre commented on ml-agents/setup.py:
```diff
@@ -0,0 +1,47 @@
+        'docopt',
+        'pyyaml',
+        'protobuf>=3.6.0',
+        'grpcio>=1.11.0'],
```
The requirements are changed no ? 
```
tensorflow==1.7.1
Pillow>=4.2.1
matplotlib
numpy>=1.11.0
jupyter
pytest>=3.2.2
docopt
pyyaml
protobuf==3.5.2
grpcio==1.11.0
```
I am worried for grpcio and protobuf since these libraries are still quite unstable

vincentpierre replied:
I am pretty sure it is not going to work with later versions of proto

dericp replied:
Good point, I'll change the setup file.

awjuliani commented on notebooks/getting-started.ipynb:
```diff
@@ -15,15 +15,13 @@
     "### 1. Set environment parameters\n",
     "\n",
-    "Be sure to set `env_name` to the name of the Unity environment file you want to launch. Ensure that the environment build is in the `python/` directory."
+    "Be sure to set `env_name` to the name of the Unity environment file you want to launch. Ensure that the environment build is in the ML-Agents SDK directory."
```
Shouldn't it be built to the root of the repo?

dericp replied:
Ah, sorry, I should either rename that to ML-Agents Toolkit or "root directory."
